xdsl/parser/attribute_parser.py

```diff
@@ -643,20 +645,98 @@ def _parse_optional_builtin_parametrized_attr(self) -> Attribute | None:
+            self.raise_error("Hex string in denseAttr is invalid")
+
+        # Use struct builtin package for unpacking f32, f64
+        format_str: str = ""
+        match element_type:
+            case Float32Type():
+                chunk_size = 4
+                format_str = "@f"  # @ in format string implies native endianess
+            case Float64Type():
+                chunk_size = 8
+                format_str = "@d"
+            case IntegerType():
+                if element_type.width.data % 8 != 0:
+                    self.raise_error(
+                        "Hex strings for dense literals only support integer types that are a multiple of 8 bits"
+                    )
+                chunk_size = element_type.width.data // 8
+            case _:
+                self.raise_error(
+                    "Hex strings for dense literals are only supported for int, f32 and f64 types"
+                )
+        num_chunks = len(byte_list) // chunk_size
+
+        data_values: list[int] | list[float] = []
+
+        if isa(element_type, Float32Type | Float64Type):
+            for i in range(num_chunks):
+                parsed_float = struct.unpack(
+                    format_str, byte_list[i * chunk_size : (i + 1) * chunk_size]
+                )
+                data_values.append(*parsed_float)
+        # Use int for unpacking IntegerType
+        else:
+            for i in range(num_chunks):
+                parsed_int = int.from_bytes(
+                    byte_list[i * chunk_size : (i + 1) * chunk_size],
+                    sys.byteorder,
+                    signed=True,
+                )
+                data_values.append(parsed_int)
+        # Splat attribute case, same value everywhere
+        if chunk_size == len(byte_list) and type.get_shape() != (1,):
+            return data_values, []
+        return data_values, [num_chunks]
+
     def _parse_builtin_dense_attr(self, _name: Span) -> DenseIntOrFPElementsAttr:
         self.parse_punctuation("<", " in dense attribute")
 
         # The flatten list of elements
-        values: list[AttrParser._TensorLiteralElement]
+        values: list[AttrParser._TensorLiteralElement] | list[int] | list[float]
         # The dense shape.
         # If it is `None`, then there is no values.
         # If it is `[]`, then this is a splat attribute, meaning it has the same
         # value everywhere.
         shape: list[int] | None
+        hex_string: str | None = None
         if self._current_token.text == ">":
             values, shape = [], None
         else:
-            values, shape = self._parse_tensor_literal()
+            hex_string = self.parse_optional_str_literal()
+            if hex_string is not None:
+                # Can not determine values without type yet
+                values, shape = [], None
+            else:
+                # Expect a tensor literal instead
+                values, shape = self._parse_tensor_literal()
```

Comment:
Why not put the str literal logic inside the `_parse_tensor_literal` code? Do you know how MLIR structures this parsing logic?

Reply:
I haven't looked at the parser code of MLIR tbh, I just parsed some denseAttrs in compiler explorer and took a look at what was produced. The MLIR codebase scares me a bit haha, but it seems that they put it inside the TensorLiteralParser?
https://github.com/llvm/llvm-project/blob/d85df3f2d6e8687c44e6802dcc0e59c14ff32c9b/mlir/lib/AsmParser/AttributeParser.cpp#L456